File: app/src/main/java/in/artsaf/seriesapp/data/api/SeriesApi.java
package in.artsaf.seriesapp.data.api;

import android.net.Uri;

import java.util.List;

import in.artsaf.seriesapp.dto.*;

public interface SeriesApi {
    public static class Contract {
        public static Uri serialsUrl(String baseUrl, String search) {
            Uri.Builder ub = Uri.parse(baseUrl).buildUpon().appendPath("serial");
            if (search != null) {
                ub.appendQueryParameter("search", search);
            }
            return ub.build();
        }

        public static Uri seasonsUrl(String baseUrl, String serialName) {
            return Uri.parse(baseUrl).buildUpon()
                    .appendPath("serial")
                    .appendPath(serialName)
                    .build();
        }

        public static Uri episodesUrl(String baseUrl) {
            return Uri.parse(baseUrl).buildUpon()
                    .appendEncodedPath("episodes/parse")
                    .build();
        }
    }

    List<Serial> serials(String search);

    List<Season> seasons(String serialName);

    Playlist episodes(String seasonHtml);
}

File: app/src/main/java/com/artsafin/seriesapp/data/api/SeriesApi.kt
package com.artsafin.seriesapp.data.api

import android.net.Uri

import com.artsafin.seriesapp.dto.*

interface SeriesApi {
    object Contract {
        fun serialsUrl(baseUrl: String, search: String?): Uri {
            val ub = Uri.parse(baseUrl).buildUpon().appendPath("serial")
            if (search != null) {
                ub.appendQueryParameter("search", search)
            }
            return ub.build()
        }

        fun seasonsUrl(baseUrl: String, serialName: String): Uri {
            return Uri.parse(baseUrl).buildUpon().appendPath("serial").appendPath(serialName).build()
        }

        fun episodesUrl(baseUrl: String): Uri {
            return Uri.parse(baseUrl).buildUpon().appendEncodedPath("episodes/parse").build()
        }
    }

    fun serials(search: String?): List<Serial>?

    fun seasons(serialName: String): List<Season>?

    fun episodes(season: Season, seasonHtml: String): Playlist?
}
